### DISCUSSION TITLE
Replacing Helm configuration with CUE?

### DISCUSSION THREAD

**Discussion by myitcv** [cue-lang/cue#1159] [Ideas]
_Relaying a conversation from Slack here for greater visibility_

_Original message from @Pothulapati_

Hello Everyone, I'm Tarun, One of the maintainers of Linkerd. I have recently been looking at cue, and totally loved the demos that I've seen. So, Currently Linkerd provides to ways to install it i.e  `CLI`  and  `Helm`. For this to work, The config is stored in a  [Helm Chart](https://github.com/linkerd/linkerd2/tree/main/charts/linkerd2), and the  [CLI acts like a wrapper around the chart](https://github.com/linkerd/linkerd2/blob/main/cli/cmd/install.go#L309). This has all the obvious problems that are present with Helm i.e All k8s options should be templated in the charts(which are endless). etc. I'm thinking what if we could replace the Helm configuration with that of Cue, but still be able to support the current Helm and the CLI ways of installing it. I also have seen that, This is common problem in the cloud native ecosystem.

So, My questions are  

-   Are there any projects that do this? Does not have to be the same problem. but some project that uses k8s, etc?
-   Any resources of examples or best practices that I can start with?
-   Also, Is this a right way to solve this? Do you have any suggestions/feedback for a problem like this?

Thanks! 🙂

**Comment by myitcv** [comment]
_Response from @verdverm_

There are a few projects / tools out there, haven't seen anything stable or prod ready. A lot of people are trying to replace Yaml / Helm with CUE.

The big thing I think about w.r.t. k8s is that we want some sort of "common" or "standard" k8s resource schema. `cue import` can be used to generate the schema from the Go structs, but there are validation gaps. Others are writing the main ones from scratch, but this seems like it could lead to everyone having their own and not being able to interop with eachother. I think the solution here is to have an official generated schema with the gaps filled.

Regarding replacing Helm more specifically, there are some features that don't necessarily work well with vanilla CUE. In particular I'm thinking about Chart versioning & incluster management, the resource ordering (i.e. what gets deployed first vs. last), checking what has actually changed and then only applying updates.

I suspect it would be easier to build a tool in Go that leverages CUE rather than trying to accomplish this in pure CUE. I came to this opinion after working on [https://github.com/hofstadter-io/cuelm](https://github.com/hofstadter-io/cuelm) and [https://github.com/hofstadter-io/c8s](https://github.com/hofstadter-io/c8s)  
I'd still like to work the pure CUE as a thought experiment and project to find where we can improve CUE to make the pure implementation a reality

**Reply by myitcv** [reply]
_Response from @morlay_


Here a tool to replace helm, too.  [https://github.com/octohelm/cuemod](https://github.com/octohelm/cuemod)

Helm supported by a tag  `@translate("helm")`  
[https://github.com/octohelm/cuemod#helm](https://github.com/octohelm/cuemod#helm)

I implements a go modules like to manage dependency.

Instead of cue official import, I add a middleware named extractor  [https://github.com/octohelm/cuemod/tree/main/pkg/extractor](https://github.com/octohelm/cuemod/tree/main/pkg/extractor) to support convert non-cue codes, like go, helm charts, jsonnet, and crds.

I will not continue use Helm,  
CUE will be better for me to manage huge components, be more composable.

So I prefer to use other tools templates in CUE.  
and final release like Tanka with nested resource objects.

Btw, if still want to support for HELM,  
we could convert all CUEs in to single CUE file  

```
# will bundle to one single cue file  
cuem eval -o ./kube.single-file.cue ./kube.cue
```

then we could create some helm plugin, to render cue with values to multi yamls.

**Reply by myitcv** [reply]
_Response from @lmickh_


Seems like a good start. My biggest problem right now is basically just trying to figure out a way to get the results of `helm template` out to jsonnet or cue.

**Reply by myitcv** [reply]
_Response from @verdverm_

I had thought about writing a parser for helm templates (you can get text/template ast in Go) but then punted on it due to the complexity. As I under stand it `helm template` renders everything and you lose the logic and organization of the templates (which is really where CUE replaces helm (text/template) with something much better). To "recover" that, the idea of `anti-unification` could be used. That is a heuristic process and is yet to be implemented afaik. I move helm by hand. It's a good excercise and can catch unknown bugs & errors along the way!

**Reply by myitcv** [reply]
_Response from @morlay_

I tried too.  
it is impossible to convert `helm template` to `CUE` directly.  
Finally, I just keep `helm template` but convert `values.yaml` to `CUE` for type constraints.Tanka choose similar way, the added a native function `helmTemplate` for `jsonnet`  
[https://github.com/grafana/tanka/blob/main/pkg/helm/jsonnet.go#L34](https://github.com/grafana/tanka/blob/main/pkg/helm/jsonnet.go#L34)

**Reply by myitcv** [reply]
_Response from @verdverm_

I think the real opportunity is in using CUE rather than Helm templates. Text interpolation of a language scoped by invisible chars is just crazy!

**Reply by myitcv** [reply]
_Response from @morlay_

I agree.  
But it is a long journey from helm template to CUE.For new project, using CUE exactly. but for old ones and open sources, It is hard to migrate all to CUE. so I prefer to extend CUE to render helm template in CUE.

**Reply by myitcv** [reply]
_Response from @leoluk_

We published our tooling here:  [https://github.com/monogon-dev/infralib/tree/main/k8s](https://github.com/monogon-dev/infralib/tree/main/k8s)

This includes ordering

**Reply by myitcv** [reply]
_Response from @verdverm_

@leoluk how are you dealing with the `port: _` issue when importing k8s schemas from Go? Didn't see a `cue.mod` dir, but maybe I missed it?

**Reply by leoluk** [reply]
There is a cue.mod directory in there

**Reply by verdverm** [reply]
https://github.com/monogon-dev/infralib/blob/0a3b5bc2cbc0e85ecc43b52c2e67fb323310498e/k8s/cue.mod/gen/k8s.io/apimachinery/pkg/util/intstr/intstr_go_gen.cue#L19

This line is used for ports in many places and essentially means they are not validated

**Reply by leoluk** [reply]
Ah, yes. We just accept that while waiting for the generator to improve

**Reply by verdverm** [reply]
I don't believe there is an improvement that can change that. The core issue is that the Go type has a custom marshal function. CUE always turns types with custom marshal functions into `_` because one cannot generally know the intent of the custom encoding and how it maps into CUE. I'd imagine there is an eventual path to enabling Go struct tags for CUE to specify these mappings, however it may not be easily accepted into upstream projects.

The "fix" is that users have to supply the correct constraints in the `cue.mod/usr/...` which is merged (unified?) with the `cue.mod/gen/...`. So given kubernetes and that this will remain there, my thought was to have an official import + `cue.mod/usr` repository. This would mean that everyone can contribute to keeping it correct and up to date, rather than everyone having to do this independently (and likely being unaware of the issue)

**Reply by myitcv** [reply]
One option here is for the owners of `k8s.io/apimachinery` to own and maintain a CUE module that represents the CUE code generated from the Go code along with the fix that @verdverm mentioned. This obviates the need for CUE users of this package to run `cue get go` (which will become [`cue import go`](https://github.com/cue-lang/cue/issues/1036)) _and_ maintain such a "fix". 

Relating this to the [package management proposal](https://github.com/cue-lang/cue/issues/851), this also touches on how the Go and CUE packages would have to be co-versioned in some way (because one is essentially a functional transformation of the other).

**Reply by verdverm** [reply]
They have some interesting requirements for their OpenAPI spec & generation https://github.com/kubernetes/kube-openapi

**Reply by luolong** [reply]
To me, the most interesting aspect of Helm is that it has a concept of "Helm repository" for Helm chart releases.
The fact that I can install a particular version of a helm chart (release) and apply environemnt specific configuration (kustomizations) to it, is the main attraction of Helm.

Now, if there was a way to package K8s app manifest bundles in some form of _"CUE release packages"_ and then apply them to my environment along with the environment specific overrides, that would be awesome and I would definitely jump ship right away.

I would not really care all that much if the _"CUE release package"_ format is not compatible with helm templates -- it doesn't need to be. Just give me tooling that gives me similar functionality to `helm isntall ...` and I happily re-write my charts in CUE :)

**Comment by helderco** [comment]
I've also been looking for a solution.

I've been thinking about https://kubevela.io/ for my apps. You can use CUE to create the components and traits, then simply put the Application manifest in the chart, and send the chart values into kubevela's properties. 

This way you still have composability, no need for templating, but still support releases. Sort of an release manager for OAM applications.
